src/query.js

```diff
@@ -1,242 +1,393 @@
+ * away.
  */
 class Query {
   /**
-   * User-supplied function to set up an individual disjoint path. Per-path
-   * query state should be held in this function's closure.
-   * @typedef {makePath} function
+   * User-supplied function to set up a query over an individual disjoint path.
+   * Per-path query state should be held in this function's closure.
+   * @typedef {makePathQuery} function
    * @param {number} pathNum - Numeric index from zero to numPaths - 1
    * @returns {queryFunc} - Function to call on each peer in the query
    */
 
   /**
    * Query function.
+   *
+   * The query function should return an object with
+   *   value: any Object - it will be yielded by the Query.run() method.
+   *   closerPeers: an array of PeerInfo objects - if supplied, Query.run()
+   *     will queue up those peers to be queried next.
+   *   pathComplete: if true, no more queries will be made on this path.
+   *   queryComplete: if true, no more queries will be made and queries
+   *     on all other paths will be stopped.
+   *
    * @typedef {queryFunc} function
-   * @param {PeerId} next - Peer to query
-   * @param {function(Error, Object)} callback - Query result callback
+   * @param {PeerId} peer - Peer to query
+   * @returns {closerPeers: Array<PeerInfo>, value: Object, pathComplete: bool, queryComplete: bool}
    */
 
   /**
-   * Create a new query. The makePath function is called once per disjoint path, so that per-path
-   * variables can be created in that scope. makePath then returns the actual query function (queryFunc) to
+   * Create a new query. The makePathQuery function is called once per disjoint path, so that per-path
+   * variables can be created in that scope. makePathQuery then returns the actual query function (queryFunc) to
    * use when on that path.
    *
    * @param {DHT} dht - DHT instance
    * @param {Buffer} key
-   * @param {makePath} makePath - Called to set up each disjoint path. Must return the query function.
+   * @param {makePathQuery} makePathQuery - Called to set up each disjoint path. Must return the query function.
    */
-  constructor (dht, key, makePath) {
+  constructor (dht, key, makePathQuery) {
     this.dht = dht
     this.key = key
-    this.makePath = makePath
-    this.concurrency = c.ALPHA
+    this.makePathQuery = makePathQuery
     this._log = utils.logger(this.dht.peerInfo.id, 'query:' + mh.toB58String(key))
+    this.running = false
+  }
+
+  /**
+   * Called when the Query starts.
+   */
+  _onStart () {
+    this.running = true
+    this.dht._queryManager.started(this)
+    this._log('run:start')
+  }
+
+  /**
+   * Stop the Query if it is running. This should be called after run() if the
+   * Query hasn't completed (eg if the caller only needs a limited number of
+   * results) so that the Query doesn't continue traversing the DHT
+   * unnecessarily.
+   */
+  stop () {
+    this.run && this.run.stop()
```

Comment:
this seems like a mistake (as I elsewhere noted about `this.run`). I would expect `this.run` to always be truthy because `run` is defined on the class